### ISSUE TITLE
Changing the Features order in the input data gives different results

### ISSUE THREAD

**Issue by Gunjitbedi** [dreamquark-ai/tabnet#316]
Hi, 

I have trained a regression model with some 4 features. But if I change the order of the features in the input csv data the results come out to be very different.

For eg.
Column order
X1,X2,X3,X4 -> rmse 40
X1,X4,X3,X2 - > rmse 35

**Comment by Optimox** [comment]
Hello,

Do you mean, « when changing the order of my columns before training from scratch my scores are different » or « after training a model, if I change the order of my columns my scores are different »?

The second option is totally expected, and you should never change the order of your columns during inference, you should keep the same order as during training.

If it’s the first option, this could be due to random seed effect (changing columns order is equivalent to changing the seed). In that case you should check the differences in scores when changing the random seed and see if they are of the same magnitude.

**Comment by Optimox** [comment]
@Gunjitbedi could you please clarify your question?

**Comment by Gunjitbedi** [comment]
@Optimox The first option.

To expand a little In my CSV data I have some features and the target variable.

While model training I have already fixed the seed so that the results/metric scores(RMSE in my case) are repeatable on the same dataset and hyperparameters.

However, if in the training data (CSV File) if I change the order of the columns/features and make NO change in my training code then I get a completely different metric score which is a little difficult for me to understand. 

How is the order of the features/columns in the CSV data influencing how the model would get trained and so affecting the metric score?

I hope this explains.

**Comment by Optimox** [comment]
Well, fixing the seed ensures that all random effects / parameters will be the same if you rerun the same code.

This is especially true for the random weights you'll start with, imagine you had a simple linear model with 4 weights `w1, w2, w3, w4` then at first iteration your model outputs for ordered input `y=w1*x1+w2*x2+w3*x3+w4*x4` then you'll perform gradient descent etc... If you reverse the order of inputs with the same seed then you'll get `y=w1*x4+w2*x3+w3*x2+w4*x1`. As you can see, what you did is exactly similar to changing the random seed.

So the differences in scores you see are coming from changes in random seed, if they are unreasonably high, this means that your model is not really converging and that random seed plays a huge role in your problem, which is not a good thing. To avoid that you should maybe try more epochs  or decaying your learning rate so that everything converges smoothly, which will make the impact of random seed smaller.

Hope it helps.

Best

**Comment by Optimox** [comment]
I'm closing this feel free to reopen if you have more information to share
